### DISCUSSION TITLE
Psycopg3 installation with --no-binary on MacOS

### DISCUSSION THREAD

**Discussion by khushboovashi** [psycopg/psycopg#433] [General]
We (pgAdmin) install psycopg with --no-binary option and then link the dependencies like libpq/openssl against our build of libpq/openssl etc. This is working fine with psycopg2 but not working with psycopg3.

As per the source code, psycopg loads the libraries from ctypes and it will only work if the libraries reside in the default location /usr/local/lib for MacOS. Is this correct ?

We need to create the pgAdmin installer in a way that all the dependencies should be in AppBundle itself. AppBundles are designed to be completely self-contained and runnable from anywhere. 

Is there any way to do it?

**Comment by dvarrazzo** [comment]
Hello @khushboovashi 

> We (pgAdmin) install psycopg with --no-binary option and then link the dependencies like libpq/openssl against our build of libpq/openssl etc. This is working fine with psycopg2 but not working with psycopg3.

The most similar installation of `psycopg2 --no-binary` is installing `psycopg[c]`. This will create two compiled C extension (like psycopg2 does installing from source).

```
$ ls -1 psycopg_c/psycopg_c/*.so
psycopg_c/psycopg_c/pq.cpython-38-x86_64-linux-gnu.so
psycopg_c/psycopg_c/_psycopg.cpython-38-x86_64-linux-gnu.so
```

which you can treat, for linking purposes, like you treat `_psycopg.*.so` in psycopg2.

> As per the source code, psycopg loads the libraries from ctypes and it will only work if the libraries reside in the default location /usr/local/lib for MacOS. Is this correct ?

I don't know if it's correct: I don't know ctypes so well, especially its behaviour on macOS. Maybe there are ways to tweak its libraries search path, such as, or similar to, setting `LD_LIBRARY_PATH`. Psycopg doesn't mess with it. But I think that installing `psycopg[c]` you will sidestep the use of ctypes.

> We need to create the pgAdmin installer in a way that all the dependencies should be in AppBundle itself. AppBundles are designed to be completely self-contained and runnable from anywhere.
> 
> Is there any way to do it?

I don't think there is anything in psycopg actively stopping you from doing so. I suspect that starting from a `psycopg[c]` installation you will find a more familiar situation to what you know from psycopg2.
